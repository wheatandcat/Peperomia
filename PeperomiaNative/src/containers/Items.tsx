import { SQLite } from "expo-sqlite";
import React, { createContext, Component } from "react";
import { db } from "../lib/db";
import { select as selectItems, Item } from "../lib/db/item";
import {
  selectByItemId as selectItemDetailByItemId,
  ItemDetail
} from "../lib/db/itemDetail";

const Context = createContext({});
const { Provider } = Context;

interface ItemAbout {
  itemId: number;
  about: string;
}

interface Props {}

interface State {
  loading: boolean;
  items: Item[];
  about: ItemAbout[];
}

export default class extends Component<Props, State> {
  state = {
    loading: true,
    items: [],
    about: []
  };

  componentDidMount() {
    this.getData();
  }

  getData = () => {
    db.transaction((tx: SQLite.Transaction) => {
      selectItems(tx, this.setItems);
    });
  };

  setItems = (data: any, error: any) => {
    if (error) {
      return;
    }
    this.setState({
      items: data,
      about: []
    });

    data.map((val: Item) => {
      db.transaction((tx: SQLite.Transaction) => {
        selectItemDetailByItemId(tx, String(val.id), this.setItemsDetail);
      });
    });
  };

  setItemsDetail = (data: any, error: any) => {
    if (error || !data || data.length === 0) {
      return;
    }

    const names = data.map((val: ItemDetail) => val.title).join("â†’");
    const itemId = data[0].itemId;
    const about = [
      ...this.state.about,
      {
        itemId: itemId,
        about: names
      }
    ];

    this.setState({
      about,
      loading: false
    });
  };

  render() {
    return (
      <Provider
        value={{
          items: this.state.items,
          about: this.state.about,
          refreshData: this.getData,
          itemsLoading: this.state.loading
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
