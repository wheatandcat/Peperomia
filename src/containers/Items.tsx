import React, {
  FC,
  createContext,
  memo,
  useState,
  useCallback,
  useContext,
} from 'react';
import { SelectItem } from 'domain/item';
import { SelectCalendar } from 'domain/calendar';
import { SelectItemDetail } from 'domain/itemDetail';
import { getItems } from 'lib/item';
import { getItemDetails } from 'lib/itemDetail';
import { getCalendars } from 'lib/calendar';
import { useDidMount } from 'hooks/index';
import { UID } from 'domain/user';
import { Context as AuthContext } from './Auth';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

type ItemAbout = {
  itemId: number | string;
  about: string;
};

type Props = {};

type State = {
  loading: boolean;
  items: SelectItem[];
  itemDetails: SelectItemDetail[];
  about: ItemAbout[];
  calendars: SelectCalendar[];
};

export type ContextProps = Partial<
  Pick<State, 'items' | 'itemDetails' | 'about' | 'calendars'> & {
    refreshData: (targetUid?: UID | false) => void;
    itemsLoading: boolean;
  }
>;

const initialState = () => ({
  loading: true,
  calendarsLoading: true,
  items: [],
  itemDetails: [],
  about: [],
  calendars: [],
});

const Connected: FC<Props> = memo((props) => {
  const [state, setState] = useState<State>(initialState());
  const { uid } = useContext(AuthContext);

  const setItemsDetail = useCallback(
    (data: SelectItemDetail[]) => {
      const names = data.map((val: SelectItemDetail) => val.title).join(' â†’ ');
      const itemId = data[0].itemId;

      const about = [
        ...state.about,
        {
          itemId,
          about: names,
        },
      ];

      setState((s) => ({
        ...s,
        about,
        loading: false,
        itemDetails: [...state.itemDetails, ...data],
      }));
    },
    [state]
  );

  const getData = useCallback(
    async (targetUid: UID | false = false) => {
      setState((s) => ({
        ...s,
        loading: true,
      }));

      let uid2 = targetUid === false ? uid : targetUid;

      const items = await getItems(uid2);

      setState((s) => ({
        ...s,
        items: items,
        about: [],
      }));

      items.map(async (val) => {
        const itemDetails = await getItemDetails(uid2, String(val.id));

        if (!itemDetails || itemDetails.length === 0) {
          return;
        }

        setItemsDetail(itemDetails);
      });

      const calendars = await getCalendars(uid2);

      setState((s) => ({
        ...s,
        calendars,
        loading: false,
      }));
    },
    [setItemsDetail, uid]
  );

  useDidMount(() => {
    getData();
  });

  return (
    <Provider
      value={{
        items: state.items,
        itemDetails: state.itemDetails,
        calendars: state.calendars,
        about: state.about,
        refreshData: getData,
        itemsLoading: state.loading,
      }}
    >
      {props.children}
    </Provider>
  );
});

export default Connected;

export const useItems = () => useContext(Context);
export const Consumer = Context.Consumer;
