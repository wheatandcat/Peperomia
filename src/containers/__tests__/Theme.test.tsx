import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Items from '../Items';

jest.mock('react-native-appearance', () => {
  const appearance = jest.requireActual('react-native-appearance');
  return {
    ...appearance,
    useColorScheme: () => jest.fn(),
  };
});

describe('containers/Items.tsx', () => {
  let wrapper: ShallowWrapper;

  const propsData: any = () => ({
    children: <div />,
  });

  beforeEach(() => {
    wrapper = shallow(<Items {...propsData()} />);
  });

  it('正常に表示されている', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
