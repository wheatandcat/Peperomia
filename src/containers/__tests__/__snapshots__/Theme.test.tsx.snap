// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Items.tsx 正常に表示されている 1`] = `
<ContextProvider
  value={
    Object {
      "about": Array [],
      "calendars": Array [],
      "itemDetails": Array [],
      "items": Array [],
      "itemsLoading": true,
      "refreshData": [Function],
    }
  }
>
  <div>
    hi
  </div>
</ContextProvider>
`;
